internal/db/schema/migrations/oss/postgres/28/01_prior_session_trigger.up.sql

```diff
@@ -0,0 +1,117 @@
+        -- If the last state is not terminated for this session...
+        if (select state from session_state where session_id in
+            (select session_id from state_counts where count = 3)
+            and session_id=target.session_id order by start_time desc limit 1)
+            != 'terminated' then
+            -- See if terminated appears; if so, prune back to it
+            if exists(select * from session_state where session_id=target.session_id and state='terminated')then
+                --Then we find the terminated record and timestamp and delete those that came before
+                delete from session_state where session_id = target.session_id and previous_end_time >(
+                    select previous_end_time from session_state where session_id=target.session_id and state='terminated');
+                -- Then remove terminated end time
+                update session_state set end_time=NULL where session_id=target.session_id and state='terminated';
+            end if;
+        end if;
+        -- Check for PCA case; if last state is not cancelled but canceling appears in the states
+        if (select state from session_state where session_id in
+            (select session_id from state_counts where count = 3)
+            and session_id=target.session_id order by start_time desc limit 1)
+            != 'canceling' then
+            -- See if canceling appears; if so, prune back to it
+            if exists(select * from session_state where session_id=target.session_id and state='canceling')then
+                --Then we find the canceling record and timestamp and delete those that came before
+                delete from session_state where session_id = target.session_id and previous_end_time >(
+                    select previous_end_time from session_state where session_id=target.session_id and state='canceling');
+                -- Then remove canceling end time
+                update session_state set end_time=NULL where session_id=target.session_id and state='canceling';
+            end if;
+        end if;
+    end loop;
+end; $$;
+
+-- Update insert session state transition trigger
+drop trigger insert_session_state on session_state;
+drop function insert_session_state();
+
+create function
+    insert_session_state()
+    returns trigger
+as $$
+begin
+
+    update session_state
+    set end_time = now()
+    where session_id = new.session_id
+      and end_time is null;
+
+    if not found then
+        new.previous_end_time = null;
+        new.start_time = now();
+        new.end_time = null;
+        return new;
+    end if;
+
+    new.previous_end_time = now();
+    new.start_time = now();
+    new.end_time = null;
+
+    new.prior_state = query.state from(
+        select state from session_state where session_id=new.session_id order by start_time desc limit 1) as query;
+
+    if new.prior_state is null then
+        new.prior_state='pending';
+    end if;
```

Comment:
Can this be incorporated into the `if not found then` block starting on line 91?